qiita_ware/search.py

```diff
@@ -0,0 +1,96 @@
+from collections import Counter, defaultdict
+
+from future.utils import viewvalues, viewitems
+
+from qiita_db.study import Study
+from qiita_db.data import ProcessedData
+
+
+def count_metadata(results, meta_cols):
+    """Counts the metadata found in a search, and returns these counts
+    Parameters
+    ----------
+    results : dict of lists of list
+        results in the format returned by the qiita_db search obj
+    meta_cols : list
+        metadata column names searched for, as returned by qiita_db search obj
+    Returns
+    -------
+    fullcount : dict of dicts
+        counts for each found metadata value over all studies, in the format
+        {meta_col1: {value1: count, value2: count, ...}, ...}
+    studycount : dict of dict of dicts
+        counts for each found metadata value for each study, in the format
+        {study_id: {meta_col1: {value1: count, value2: count, ...}, ...}, ...}
+    """
+    def double_comprehension(results):
+        for samples in viewvalues(results):
+            for sample in samples:
+                yield sample
+
+    fullcount = {}
+    # rearrange all samples so that each metadata column found is its own list
+    meta_vals = zip(*double_comprehension(results))
```

Comment:
ah, I see. There really is an incredible amount of work being done with complex nested data structures in this method. It smells like there is quite a bit of redundant work. Given the inputs, returns, and the docstring, wouldn't the following suffice? It should require a lot less iteration and allocation

```python
n_cols = len(meta_cols)
studycount = {}
fullcount = {mc: defaultdict(int) for mc in meta_cols}
for study_id, meta_vals in viewitems(results):
    metadata_counts = [defaultdict(int) for i in range(n_cols)]
    for cat_idx, meta_col, meta_val in izip(range(n_cols), meta_cols, meta_vals):
        metadata_counts[cat_idx][meta_val] += 1
        fullcount[meta_col][meta_val] += 1
    studycount[study_id] = {mc: mcounts for mc, mcounts in izip(meta_cols, metadata_counts)}
```

The above is still placing extreme faith that `meta_cols` is in order with `meta_vals`, performing redundant work, and redundant representation. The latter two points are because `fullcount` really should be expressed as `reduce(add, studycount.values())`, however `dict` doesn't support that. What I'm wondering is actually if these data should be coming in as either `np.array` or `sp.sparse.spmatrix`?